src/mtype.d

```diff
@@ -6464,73 +6465,95 @@ extern (C++) final class TypeFunction : TypeNext
-                continue;
-
-            if (fparam.storageClass & (STClazy | STCout))
-            {
-                tf.purity = PUREweak;
-                break;
-            }
-            if (fparam.storageClass & STCref)
+            if (isref)
             {
                 if (t.mod & MODimmutable)
-                    continue;
+                    return PUREstrong;
                 if (t.mod & (MODconst | MODwild))
-                {
-                    tf.purity = PUREconst;
-                    continue;
-                }
-                tf.purity = PUREweak;
-                break;
+                    return PUREconst;
+                return PUREweak;
             }
 
             t = t.baseElemOf();
-            if (!t.hasPointers())
-                continue;
-            if (t.mod & MODimmutable)
-                continue;
+
+            if (!t.hasPointers() || t.mod & MODimmutable)
+                return PUREstrong;
 
             /* Accept immutable(T)[] and immutable(T)* as being strongly pure
              */
             if (t.ty == Tarray || t.ty == Tpointer)
             {
                 Type tn = t.nextOf().toBasetype();
                 if (tn.mod & MODimmutable)
-                    continue;
+                    return PUREstrong;
                 if (tn.mod & (MODconst | MODwild))
-                {
-                    tf.purity = PUREconst;
-                    continue;
-                }
+                    return PUREconst;
             }
 
             /* The rest of this is too strict; fix later.
              * For example, the only pointer members of a struct may be immutable,
              * which would maintain strong purity.
+             * (Just like for dynamic arrays and pointers above.)
              */
             if (t.mod & (MODconst | MODwild))
-            {
-                tf.purity = PUREconst;
+                return PUREconst;
+
+            /* Should catch delegates and function pointers, and fold in their purity
+             */
+            return PUREweak;
+        }
+
+        purity = PUREstrong; // assume strong until something weakens it
+
+        /* Evaluate what kind of purity based on the modifiers for the parameters
+         */
+        const dim = Parameter.dim(tf.parameters);
+        foreach (i; 0 .. dim)
+        {
+            Parameter fparam = Parameter.getNth(tf.parameters, i);
+            Type t = fparam.type;
+            if (!t)
                 continue;
+
+            if (fparam.storageClass & (STClazy | STCout))
+            {
+                purity = PUREweak;
+                break;
             }
+            switch (purityOfType((fparam.storageClass & STCref) != 0, t))
+            {
+                case PUREweak:
+                    purity = PUREweak;
+                    break;
 
-            /* Should catch delegates and function pointers, and fold in their purity
+                case PUREconst:
+                    purity = PUREconst;
```

Comment:
> This could increase purity again, couldn't it?
No, because any setting of `PUREweak` exits the loop.
> How about replacing the switch with purity = min(purity, purityOfType(isref, t))?
Because there is no reason to continue with the loop once purity is set to `PUREweak`.